### PR TITLE
KAFKA-18026: KIP-1112, ProcessorWrapper API with PAPI and partial DSL implementation

### PR DESCRIPTION
**Pull request by ableegoldman**
This PR includes the API for KIP-1112 and a partial implementation, which wraps any processors added through the PAPI and the DSL processors that are written to the topology through the ProcessorParameters#addProcessorTo method.

Further PRs will complete the implementation by converting the remaining DSL operators to using the #addProcessorTo method, and future-proof the processor writing mechanism to prevent new DSL operators from being implemented incorrectly/without the wrapper

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/Topology.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/Topology.java:717]: this is the entire implementation for PAPI apps

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractConfigurableStoreFactory.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractConfigurableStoreFactory.java:37]: side fix: we weren't configuring this with the app configs

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java:86]: this is how the DSL implementation will work -- processors are wrapped when building the DSL into a Topology, specifically in this method where they are added to the InternalTopologyBuilder.

Note this PR is only a partial implementation since not all DSL operators go through this class, some just write their processors to the InternalTopologyBuilder directly. Followup PRs will tackle converting these operators and enforcing that any new ones added in the future will have to go through this method and be wrapped

File: streams/src/main/java/org/apache/kafka/streams/TopologyConfig.java
Comment (agavra) [inline streams/src/main/java/org/apache/kafka/streams/TopologyConfig.java]: Is there a possibility of having this be available (and respected) in `StreamsConfig`? People tend to forget to properly pipe things through TopolgoyConfig, though having it only there makes it a bit more obvious where it needs to be passed in.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (agavra) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: nit: does it make sense to also include the error in the logged message in case the user swallows this up higher in the stack somewhere?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java
Comment (agavra) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java:86]: curious why we do this here instead of inside `topologyBuilder.addProcessor`

**Review by agavra — APPROVED** [review]
LGTM! (non-binding, of course)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java:86]: ah, good question! that would definitely be simpler. however, to maximize the utility of the new wrapper, we want to allow users to wrap not only the Processor returned by the ProcessorSupplier#get method, but also the StoreBuilders returned from ProcessorSupplier#stores. Unfortunately, the stores are currently extracted and connected outside of the #addProcessor method, and since they aren't attached (via #addStateStore) until after #addProcessor is called, we can't even look up and wrap the stores for a processor from within #addProcessor

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java:86]: Ultimately it might make sense to move the extraction of the stores from a ProcessorSupplier into a single InternalTopologyBuilder method (eg #addStatefulProcessor) that calls both #addProcessor and #addStateStore, and do the wrapping there. This will also help future-proof new DSL operators from being implemented incorrectly and missing the wrapper step. 

But I think it makes sense to wait until we've completed all the followup PRs for the remaining DSL operators, in case there are any weird edge cases we haven't thought of/seen yet. Then we can do one final PR to clean up and future-proof the wrapping process

File: streams/src/main/java/org/apache/kafka/streams/TopologyConfig.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/TopologyConfig.java]: ah, good catch -- this comment is actually out of date, and I did end up having to add the config to StreamsConfig as well.

I definitely share your concerns about people forgetting to properly pipe things in to the Topology/StreamsBuilder through TopologyConfig though. I recently filed [KAFKA-18053](https://issues.apache.org/jira/browse/KAFKA-18053) for exactly this problem, with one subtask covering a way to improve the situation without a KIP by detecting and notifying users of a misconfiguration. Ideally we can include this fix in 4.0 as well

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java]: good point

File: streams/src/main/java/org/apache/kafka/streams/TopologyConfig.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/TopologyConfig.java:80]: Thanks for adding the docs here, would be helpful to users.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:91]: nit: " that"?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:2281]: Just checking if this is correct? `wrapFixedKeyProcessorSupplier` func itself calls `return ProcessorWrapper.asWrappedFixedKey(processorSupplier);` already, and hence seems a circular calling?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:2289]: Similar here.

**Review by guozhangwang — COMMENTED** [review]
If all the unit tests are already in place, could you update the description as well to remove the comment that it is still WIP?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:91]: space is added at the end of the line above (admittedly it's hard to see this way)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:2281]: Ah, yes, it's correct but I see how it's confusing -- basically the non-static `#wrapFixedKeyProcessorSupplier` is what does the actual wrapping via the configured ProcessorWrapper instance, whereas the static `ProcessorWrapper#asWrapped` method is what converts a regular ProcessorSupplier into the `WrappedProcessorSupplier` subclass that we use as a marker interface to distinguish between processors that are and aren't wrapped yet.

I will add a comment to the code to clarify for future readers, but perhaps it's also worth choosing a better name for the static `#asWrapped` utility methods. Maybe `#toWrappedProcessorSupplier` or something like that?

@guozhangwang @agavra thoughts?

(btw I am going to merge the PR as-is since we can address this in a followup PR, I'll make sure to add the code comments and/or change the name in another PR)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:2289]: see above 🙂

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:2281]: I see, yeah I think having a different name for the static utility function would be better.

Comment (rodesai) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:2281]: > Ah, yes, it's correct but I see how it's confusing -- basically the non-static #wrapFixedKeyProcessorSupplier is what does the actual wrapping via the configured ProcessorWrapper instance, whereas the static ProcessorWrapper#asWrapped method is what converts a regular ProcessorSupplier into the WrappedProcessorSupplier subclass that we use as a marker interface to distinguish between processors that are and aren't wrapped yet.

I'm not following this. `wrappedFixedKeyProcessorSupplier` is already expected to return an instance that implements `WrappedProcessorSupplier` - so why does `InternalTopologyBuilder` also need to enclose it in a `WrappedProcessorSupplierImpl`? Do we explicitly check for the `WrappedProcessorSupplierImpl` class elsewhere?

Comment (agavra) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:2281]: +1 to Rohan's question, I don't think we need to do this here since the `WrappedProcessorSupplier` already returns the right type.

Just on the naming discussion, though, I think `markWrapped` could be a good name for the static method.

**Comment by cadonna** [comment]
@ableegoldman May I ask why you merged this PR without any approval from a committer excluding yourself?
At least, I was not able to find the approval.

**Pull request merged**
